File: problems11/lab11_e.py
from asyncio.windows_events import NULL
from sys import setrecursionlimit
import threading
setrecursionlimit(10 ** 9)
threading.stack_size(3 * 67108864)

def main():
    inf = 10 ** 9
    #Ford-Bellman algo https://e-maxx.ru/algo/negative_cycle
    file_input, file_output = open('negcycle.in', 'r'), open('negcycle.out','w')
    n = int(file_input.readline())
    weight_map, matrix, negcycle = [0] * n, [], [None] * n
    for i in range(n):
        current = list(map(int, file_input.readline().split()))
        # concurrent = []
        for j in range(n):
            if current[j] != inf: matrix.append([i, j, current[j]])
        # matrix[i] = concurrent
        # del concurrent
    flag = None
    for i in range(n):
        for j in range(len(matrix)):
            if weight_map[matrix[j][1]] > matrix[j][2] + weight_map[matrix[j][0]]:
                negcycle[matrix[j][1]], flag = matrix[j][:2]
                weight_map[matrix[j][1]] = matrix[j][2] + weight_map[matrix[j][0]]
        if flag is None: break
    if flag is not None:
        all_flag = flag
        for _ in range(n): all_flag = negcycle[all_flag]
        all_negcycle = []
        iter = all_flag
        while 1:
            all_negcycle.append(iter)
            iter = negcycle[iter]
            if all_flag == iter: break
        all_negcycle.append(all_flag)
        all_negcycle = all_negcycle[::-1]
        print('YES', len(all_negcycle), sep = '\n', file=file_output)
        print(*[vertex + 1 for vertex in all_negcycle], file=file_output)
    else: print('NO', file=file_output)
    file_output.close()

thread = threading.Thread(target=main)
thread.start()

File: problems11/lab11_e.cpp
#include <iostream>
#include <fstream>
#include <vector>
const long long inf = 1e9;
struct Edge {
    long long vertex_first, vertex_second, weight;
};

int main() {
    std::ifstream fin("negcycle.in");
    std::ofstream fout("negcycle.out");
    long long n;
    fin >> n;
    std::vector<Edge> matrix(n);
    std::vector<long long> negcycle(n, -1);
    std::vector<long long> weight_map(n, inf);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int weight;
            fin >> weight;
            matrix.push_back({i, j, weight});
        }
    }
    long long flag;
    for (int _ = 0; _ < n; _++) {
        flag = -1;
        for (Edge j : matrix) {
            if (weight_map[j.vertex_second] > weight_map[j.vertex_first] + j.weight) {
                weight_map[j.vertex_second] = weight_map[j.vertex_first] + j.weight;
                negcycle[j.vertex_second] = j.vertex_first;
                flag = j.vertex_second;
            }
        }
    }
    if (flag == -1) {fout << "NO";}
    else {
        std::vector<long long> all_negcycle;
        long long all_flag = flag;
        for (long long i = 0; i < n; i++) {all_flag = negcycle[all_flag];}
        for (long long i = all_flag; i != all_flag || all_negcycle.empty(); i = negcycle[i]) {all_negcycle.push_back(i);}
        all_negcycle.push_back(all_flag);
        fout << "YES\n" << all_negcycle.size() << "\n";
        for (long long i = all_negcycle.size() - 1; i >= 0; i--) {fout << all_negcycle[i] + 1 << " ";}
    }
    return 0;
}